assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js

```diff
@@ -104,24 +104,37 @@ function useUploadMedia({ media, pagingNum, mediaType, fetchMedia, setMedia }) {
       }
 
       try {
-        const uploadingFiles = await Promise.all(
+        const uploadedFiles = await Promise.all(
           localFiles.map(async (localFile) => ({
             ...localFile,
-            fileUploaded: await uploadFile(localFile.file),
+            fileUploaded: getResourceFromAttachment(
+              await uploadFile(localFile.file)
+            ),
           }))
         );
 
         setIsUploading(false);
 
         if (onUploadedFile) {
-          uploadingFiles.forEach(({ element, fileUploaded }) => {
+          uploadedFiles.forEach(({ element, fileUploaded }) => {
             onUploadedFile({
-              resource: getResourceFromAttachment(fileUploaded),
+              resource: fileUploaded,
               element,
             });
           });
         }
-        fetchMedia({ pagingNum, mediaType }, setMedia);
+
+        const uploadedFilesMap = new Map(
+          uploadedFiles.map(({ localResource, fileUploaded }) => [
+            localResource,
+            fileUploaded,
+          ])
+        );
+        setMedia({
```

Comment:
Seems like reshuffling the whole array of media elements is a lot. I think we already have `updateMediaElement`. Can we use it here?

Reply:
As it is today, the reducer for updateMediaElement is also O(n) for each media element. But we should be able to refactor it to take more than one media element at a time.

Here is an issue to address this: https://github.com/google/web-stories-wp/issues/2072